### PR TITLE
Add revision tracking to drafts, remove button, remove from new page

### PR DESCRIPTION
**Pull request by stephaniehobson**
The goal of this PR is to make the auto save drafting feature reliable.

NOT READY. But this is a big PR and I would appreciate a first pass review while I fight the edge cases.

Still to do:
- [x] Fix #todo links for revision history
- [x] Write tests

Fix Bug 1264314: Alert user if page content has changed since last draft
- revision ID is now saved when draft saved
- displays warning message if revision ID is old
- displays note if revision ID is old
  - Fix Bug 1298399: Remove ?document_saved
- query variable now includes revision ID of draft to delete
- query variable now directly consumed by javascript
- query variable new removed after draft deleted
  - Fix Bug 1268906: Tool bar buttons now trigger draft saves
- changed listener from 'key' to 'change'
  - Fix Bug 1274067: No new drafts created after discarding page
- disabled autosave after discard button clicked
  - Fix Bug 1269851: Save draft should not wait for typing break to save
- added throttled call to autosave every 500ms
- reduced debounce autosave to 500ms, to catch edits made after throttle
  - Fix Bug 1301656: Remove draft feature from new page creation
- removed draft widget from new pages
- changed data format for new translation pages to retain drafts there

Also:
- lint wiki-edit.js
- remove unused wiki-section-edit.js
- autosave now enabled by default even with saved draft
  - removed need to discard saved draft
- separated draft functionality into a new file
- new utility function to get query strings by variable name
- added some analytics instrumentation
- standardized draft storage formats for non-English pages
  - migrated old draft formats to new storage format

### CODE REVIEW

**Comment by stephaniehobson** [comment]
@schalkneethling  I'd appreciate an initial review. I'm still fighting with some edge cases and need to add tests.

@jwhitlock You know the weird rabbit warrens that are the create / edit / translate paths in kuma, can you look and see if I've left an edge case unaccounted for? Create English pages should not get drafts anymore, new translations should, edits should, and translation edits should.

If I work on this at all tomorrow it will be tests, so no hurry on the review. I have a few reviews to do myself ;)

File: kuma/static/js/wiki.js [hunk 1]
```diff
@@ -976,10 +976,40 @@
...
-    // clear out any drafts from localStorage, now that the document has been saved
-    if (typeof document_saved !== 'undefined' && document_saved) {
-      localStorage.removeItem('draft/edit' + location.pathname);
-      localStorage.removeItem('draft/edit' + location.pathname + '#save-time');
+
+    //
+    // Generates a storage key based on pathname
+    // copied from wiki-edit-draft.js because: race conditions
+    // does not need to copy logic for dealing with new translations
+    //
```
Comment (schalkneethling) [inline kuma/static/js/wiki.js]: We should probably start moving to some cross project standards for comments. In this case we probably want to do:

```
/**
 * Generates a storage key based on pathname
 * copied from wiki-edit-draft.js because: race conditions
 * does not need to copy logic for dealing with new translations
 */
```
Probably a good idea to follow the JSDoc guidelines http://usejsdoc.org/ - unless MDN already uses another JSDoc standard or tool.

Comment (stephaniehobson) [inline kuma/static/js/wiki.js]: MDN is the wild west of javascript formatting right now (I just copy what is in the file I'm working on at the time). Cross project standards would be awesome. 👍

I will try to get over my distaste for using comments that are not left aligned :P

**Comment by stephaniehobson** [comment]
Darn, legitimate test failure. That's what I get for editing Python...

File: kuma/static/js/wiki-edit-draft.js [hunk 2]
```diff
@@ -0,0 +1,428 @@
+(function ($, win, doc) {
+    'use strict';
+
+    // feature test variable
+    var supportsLocalStorage = win.mdn.features.localStorage;
+    // some knowledge about the page we're dealing with
+    var $translateForm = $('#translate-document');
+    var $form = $translateForm.length ? $translateForm : $('#wiki-page-edit');
+    var isNewTranslation = location.search.indexOf('?tolocale') > -1 ? true : false;
```
Comment (jwhitlock) [inline kuma/static/js/wiki-edit-draft.js]: Will ``tolocale`` always be the first query string parameter?

File: kuma/static/js/main.js [hunk 3]
```diff
@@ -24,6 +24,16 @@
...
 
     /*
+        Get URL Parameter
+    */
+    win.mdn.getUrlParameter = function(name) {
+        name = name.replace(/[\[]/, '\\[').replace(/[\]]/, '\\]');
+        var regex = new RegExp('[\\?&]' + name + '=([^&#]*)');
+        var results = regex.exec(location.search);
+        return results === null ? '' : decodeURIComponent(results[1].replace(/\+/g, ' '));
+    };
```
Comment (jwhitlock) [inline kuma/static/js/main.js]: JavaScript doesn't have a native method for this? 😢

File: kuma/static/js/wiki-edit-draft.js [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+    // stuff we need for drafts
+    var DRAFT_NAME;
+    var draftTimeOutId;
+    var draftAutoSaveEnabled = true;
+    var draftWait = false;
+    var draftWaitTime = 500;
+    var $revisionInputs = $('[name=current_rev]'); // can't use ID because there's a duplicate >_<
```
Comment (jwhitlock) [inline kuma/static/js/wiki-edit-draft.js]: We could try to fix this in the form / template logic...

File: kuma/static/js/wiki-edit-draft.js [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+        key = key.replace('$edit', '');
+        key = key.replace('$translate', '');
+
+        key = 'draft/edit' + key;
+
+        if (isNewTranslation) {
+            // change translation creation format to match edit format
+            // /en-US/docs/page$translate?tolocale=fr
+            // draft/edit/fr/docs/page
+            var localeCode = location.search.replace('?tolocale=', '');
```
Comment (jwhitlock) [inline kuma/static/js/wiki-edit-draft.js]: could use the new ``getUrlParameter``

File: kuma/static/js/wiki-edit-draft.js [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+                    action: 'outdated-draft-warning'
+                });
+            }
+
+            if (draftType === 'unknown') {
+                $draftOld.addClass('note');
+                draftWarning += '<div class="readable-line-length">';
+                // long line is long, but breaking it up wrecks it for the localizers
+                draftWarning += gettext('Compare this date to the latest revision date to ensure you\'re not overwriting later changes.');
+                draftWarning += ' <a href="#TODO" class="external-icon">Revision history.</a>';
```
Comment (jwhitlock) [inline kuma/static/js/wiki-edit-draft.js]: ``Revision history.`` should be localized too.  Which will make this line even longer...

File: kuma/static/js/wiki-edit-draft.js [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+                draftType = 'unknown';
+            }
+
+            if (draftType === 'old') {
+                restore = gettext('View draft.');
+                $draftOld.addClass('warning');
+                draftWarning += '<div class="readable-line-length">';
+                // long line is long, but breaking it up wrecks it for the localizers
+                draftWarning += gettext('A newer version of this article has been published since this draft was saved. You can restore the draft to view the content, but you will not be able to submit it for publishing.');
+                draftWarning += ' <a href="#TODO" class="external-icon">Revision history.</a> ';
```
Comment (jwhitlock) [inline kuma/static/js/wiki-edit-draft.js]: ``Revision history`` should be localized.  Is the ``#TODO`` for this PR?

File: kuma/static/js/wiki-edit-draft.js [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+                draftWarning += '<a href="#TODO" class="external-icon">Published version</a>';
```
Comment (jwhitlock) [inline kuma/static/js/wiki-edit-draft.js]: ``Published version`` should be localized.  Is the ``#TODO`` for this PR?

File: kuma/static/js/wiki-edit-draft.js [hunk 9]
```diff
@@ -0,0 +1,428 @@
...
+            // clone elements and populate
+            var $updateAction = $draftAction.clone().text(action);
+            var $updateTime = $draftTime.clone()
+                .attr('title', now.toISOString())
+                .text(nowString);
+            // append elements
+            $draftStatus.append(gettext('Draft') + ' ')
+                .append($updateAction)
+                .append(' ')
+                .append($updateTime);
```
Comment (jwhitlock) [inline kuma/static/js/wiki-edit-draft.js]: I'm not sure localized strings can be combined this way in all locales.  It may be better to pass the two values in ('published' and 'discarded') and construct translatable strings for each, like
```
if (action === 'published') {
   $draftStatus.append(interpolate("Draft published at %s", $updateTime))
}
```

File: kuma/static/js/wiki-edit-draft.js [hunk 10]
```diff
@@ -0,0 +1,428 @@
...
+
+    /*
+    function manualSaveDraft() {
+        var currentContent = $form.find('textarea[name=content]').val();
+        // save draft
+        saveDraft(currentContent, gettext('saved'));
+        // disable autosave when saving manually
+        enableAutoSave(false);
+    }
+    */
```
Comment (jwhitlock) [inline kuma/static/js/wiki-edit-draft.js]: Remove, if you've gotten consensus to remove manual draft saving.

**Comment by jwhitlock** [comment]
Ug, thought I was leaving a review. Sorry for the spam...

File: kuma/static/js/main.js (hunk 3 above)
Comment (stephaniehobson) [inline kuma/static/js/main.js]: i no rite?

File: kuma/static/js/wiki-edit-draft.js (hunk 2 above)
Comment (stephaniehobson) [inline kuma/static/js/wiki-edit-draft.js]: Not sure, but this should be okay if it's not.

File: kuma/static/js/wiki-edit-draft.js (hunk 4 above)
Comment (stephaniehobson) [inline kuma/static/js/wiki-edit-draft.js]: That would be nice. It broke things when I tried and it's an ancient problem so I figured it could be dealt with separately but if you want to pair on this we could try.

File: kuma/static/js/wiki-edit-draft.js (hunk 9 above)
Comment (stephaniehobson) [inline kuma/static/js/wiki-edit-draft.js]: I'm 100% sure they can't be but this is what was there before.

**Comment by stephaniehobson** [comment]
Should be down to one test failure.

**Comment by stephaniehobson** [comment]
Okay, ready for final reviews now! 

Still writing tests...

File: kuma/wiki/tests/test_views.py [hunk 11]
```diff
@@ -1761,9 +1761,10 @@ def _run_translate_edit_tests(edit_slug, edit_data, edit_doc):
                 eq_(0, len(Document.objects.filter(title=edit_data['title'] +
                                                    ' Redirect 1',
                                                    locale=foreign_locale)))
-                self.assertRedirects(response, reverse('wiki.document',
-                                                       locale=foreign_locale,
-                                                       args=[edit_doc.slug]))
+                doc_url = reverse('wiki.document', args=[edit_doc.slug], locale=foreign_locale)
+                params = {'rev_saved': ''}
+                doc_url = '%s?%s' % (doc_url, urlencode(xparams))
```
Comment (jwhitlock) [inline kuma/wiki/tests/test_views.py]: This should be ``params`` not ``xparams``, as caught by [flake8](https://travis-ci.org/mozilla/kuma/jobs/228551964#L382).  It's also code that isn't called during tests, but we agreed to postpone cleanup until after this PR is merged.

**Comment by stephaniehobson** [comment]
@schalkneethling Did you want to take another look at this before it goes live? The code is "final" now I'm just writing tests.

**Comment by schalkneethling** [comment]
Looks good to me @stephaniehobson r+

**Comment by stephaniehobson** [comment]
I think this is good to go. @jwhitlock I think I have addressed the changes you requested.

**Comment by jwhitlock** [comment]
😄 will test it out Tuesday morning

**Comment by stephaniehobson** [comment]
Tests still failing :( I thought we had that beat. So much for my triumphant weekend.

File: kuma/wiki/views/edit.py [hunk 12]
```diff
@@ -245,29 +245,89 @@ def edit(request, document_slug, document_locale, revision_id=None):
...
-                        params['section'] = section_id
-                # Parameter for the document saved, so that we can delete the cached draft on load
-                params['document_saved'] = 'true'
-                url = '%s?%s' % (url, urlencode(params))
-                if not is_raw and section_id:
-                    # If a section was edited, jump to the section anchor
-                    # if we're not getting raw content.
-                    url = '%s#%s' % (url, section_id)
-                return redirect(url)
+                    orig_rev = Revision.objects.get(pk=orig_rev_id)
```
Comment (jwhitlock) [inline kuma/wiki/views/edit.py]: This rebase pulled in too much code. Lines 248-307 should be removed, and then the redirect URL code re-indented.

File: tests/functional/test_article_new.py [hunk 13]
```diff
@@ -26,9 +27,10 @@ def test_new(base_url, selenium):
     assert 'iframe' not in content
     # content edit remains
     assert 'Pumpkin' in content
-    # check contents of draft
-    draft_content = editor.draft_content(selenium.current_url)
-    assert 'Pumpkin' in draft_content
+    # wait for throttled draft function to activate (hopefully not)
+    time.sleep(0.6)
+    # check save draft not activated
+    assert not page.is_draft_container_displayed
```
Comment (jwhitlock) [inline tests/functional/test_article_new.py]: This should be ``page.editor().is_draft_container_displayed``.

File: tests/pages/regions/ckeditor.py [hunk 14]
```diff
@@ -31,33 +38,104 @@ def ready(self):
...
+    def is_draft_container_displayed(self):
+        return self.find_element(
+            *self._draft_container_locator
+        ).is_displayed()
+
+    @property
+    def is_draft_old_displayed(self):
+        return self.find_element(
+            *self._draft_old_locator
+        ).is_displayed()
```
Comment (jwhitlock) [inline tests/pages/regions/ckeditor.py]: This will raise an exception if the element doesn't exist, which is being tested in ``test_article_new``.  I'd add an import at the top:

    from selenium.common.exceptions import NoSuchElementException

and then write it like:

    def is_draft_old_displayed(self):
        try:
            return self.find_element(
                *self._draft_old_locator
            ).is_displayed()
        except NoSuchElementException:
            return False

File: tests/pages/regions/ckeditor.py [hunk 15]
```diff
@@ -31,33 +38,104 @@ def ready(self):
...
+        old_draft = self.find_element(*self._draft_old_locator)
+        return 'you will not be able to submit it' in old_draft.text
+
+    @property
+    def is_draft_restore_displayed(self):
+        try:
+            return self.find_element(
+                *self._draft_restore_locator
+            ).is_displayed()
+        except:
```
Comment (jwhitlock) [inline tests/pages/regions/ckeditor.py]: ``except:`` is too general, please make this ``except NoSuchElementException:``

File: tests/pages/regions/ckeditor.py [hunk 16]
```diff
@@ -31,33 +38,104 @@ def ready(self):
...
+
+    def draft_discarded_status(self):
+        old_draft_status = self.find_element(*self._draft_old_locator)
+        return 'Draft discarded' in old_draft_status.text
+
+    @property
+    def is_draft_status_displayed(self):
+        try:
+            return self.find_element(*self._draft_status_locator).is_displayed()
+        except:
```
Comment (jwhitlock) [inline tests/pages/regions/ckeditor.py]: ``except NoSuchElementException:`` here as well

File: tests/pages/article_new.py [hunk 17]
```diff
@@ -13,16 +14,20 @@ class NewPage(EditPage):
...
         publish_button = self.find_element(*self._save_button_locator)
         publish_button.click()
         # wait for article page to load
         published_page = ArticlePage(
             self.selenium,
             self.base_url,
-            **self.url_kwargs
+            locale=self.DEFAULT_LOCALE,
+            slug=self.DOC_SLUG,
+            # **self.url_kwargs
```
Comment (jwhitlock) [inline tests/pages/article_new.py]: This comment can be deleted

File: tests/pages/article.py [hunk 18]
```diff
@@ -83,15 +85,16 @@ def is_edit_button_displayed(self):
         return self.find_element(*self._edit_button_locator).is_displayed()
 
     def click_edit(self, signedin):
+        from pages.article_edit import EditPage
         edit_button = self.find_element(*self._edit_button_locator)
         edit_button.click()
         if (signedin):
-            from pages.article_edit import EditPage
```
Comment (jwhitlock) [inline tests/pages/article.py]: 👍 removing this import

File: tests/pages/regions/ckeditor.py [hunk 19]
```diff
@@ -31,33 +38,104 @@ def ready(self):
...
+        draft = self.selenium.execute_script(
             'return localStorage.getItem("{}");'.format(draft_name)
         )
+        if (draft is None):
+            return ''
+        else:
+            return draft
 
-    def edit_source(self):
+    def edit_source(self, content=TEST_TEXT):
```
Comment (jwhitlock) [inline tests/pages/regions/ckeditor.py]: 👍 making this an optional parameter instead of a class constant

**Comment by stephaniehobson** [comment]
Made requested changes to tests 🎉

**Comment by beanoagin** [comment]
Hi
